File: leetcode/239.滑动窗口最大值/Solution.kt
/*
 * @lc app=leetcode.cn id=239 lang=kotlin
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        if (nums == null || nums.size < 2) {
            return nums
        }
        val ans = IntArray(nums.size - k + 1)
        val deque = LinkedList<Int>()
        for (i in nums.indices) {
            while(deque.isNotEmpty() && deque.peekFirst() < (i + 1 - k)) {
                deque.removeFirst()
            }
            while(deque.isNotEmpty() && nums[deque.peekLast()] < nums[i]) {
                deque.removeLast()
            }
            deque.add(i)
            if (i >= k - 1) {
                ans[i + 1 - k] = nums[deque.peekFirst()]
            }
        }
        return ans
    }
}
// @lc code=end

File: leetcode/239.滑动窗口最大值/solution.py
#
# @lc app=leetcode.cn id=239 lang=python3
#
# [239] 滑动窗口最大值
#

# @lc code=start
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = collections.deque()
        out = []
        for i, value in enumerate(nums):
            while dq and nums[dq[-1]] < value:
                dq.pop()
            dq += i,
            if dq[0] == i - k:
                dq.popleft()
            if  i >= k - 1:
                out += nums[dq[0]],
        return out
# @lc code=end